src/lib.rs

```diff
@@ -70,103 +139,51 @@ impl<R> Volatile<R> {
     /// ## Example
     ///
     /// ```rust
-    /// use volatile::Volatile;
+    /// # extern crate core;
+    /// use volatile::VolatilePtr;
+    /// use core::ptr::NonNull;
     ///
     /// let mut value = 0u32;
     ///
-    /// let mut volatile = Volatile::new(&mut value);
+    /// let mut volatile = unsafe { VolatilePtr::new_read_write(NonNull::from(&mut value)) };
     /// volatile.write(1);
     /// assert_eq!(volatile.read(), 1);
     /// ```
-    pub const fn new(reference: R) -> Volatile<R> {
-        Volatile {
-            reference,
-            access: PhantomData,
-        }
+    pub const unsafe fn new_with_access<A>(
```

Comment:
If we were to just delete this function and only have `VolatilePtr::new_generic`, would we really lose much? It seems like you would end up writing about the same amount of code.
```rust
let access = Access {
    read: UnsafeAccess,
    write: SafeAccess,
};
let mut volatile = unsafe { VolatilePtr::new_with_access(NonNull::from(&mut val), access) };
```
vs
```rust
type Custom = Access<UnsafeAccess, SafeAccess>;
let mut volatile = unsafe { VolatilePtr::new_generic::<Custom>(NonNull::from(&mut val)) };
// Or just define inline
let mut volatile: VolatilePtr<_, Access<UnsafeAccess, SafeAccess>> = unsafe { VolatilePtr::new_generic(NonNull::from(&mut val)) };
```

It seems like having to pass the marker type by value isn't really necessary.